jetbrains-core/src/software/aws/toolkits/jetbrains/core/terminal/OpenAwsLocalTerminal.kt

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.terminal
+
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.LangDataKeys
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.project.DumbAware
+import com.intellij.util.ExceptionUtil
+import icons.TerminalIcons
+import org.jetbrains.plugins.terminal.TerminalTabState
+import org.jetbrains.plugins.terminal.TerminalView
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.jetbrains.core.credentials.AwsConnectionManager
+import software.aws.toolkits.jetbrains.core.credentials.ConnectionState
+import software.aws.toolkits.jetbrains.core.credentials.shortName
+import software.aws.toolkits.jetbrains.core.plugins.pluginIsInstalledAndEnabled
+import software.aws.toolkits.resources.message
+import software.aws.toolkits.telemetry.AwsTelemetry
+import software.aws.toolkits.telemetry.Result
+
+class OpenAwsLocalTerminal : AnAction(message("aws.terminal.action"), message("aws.terminal.action.tooltip"), TerminalIcons.OpenTerminal_13x13), DumbAware {
+
+    override fun update(e: AnActionEvent) {
+        if (!pluginIsInstalledAndEnabled("org.jetbrains.plugins.terminal")) {
+            e.presentation.isEnabled = false
+            e.presentation.isVisible = false
+            return
+        }
+        e.presentation.isEnabled = e.project?.let { AwsConnectionManager.getInstance(it) }?.isValidConnectionSettings() == true
+    }
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(LangDataKeys.PROJECT)
+        when (val state = AwsConnectionManager.getInstance(project).connectionState) {
+            is ConnectionState.ValidConnection -> {
+                val connection = state.connection
+                val credentials = try {
+                    connection.credentials.resolveCredentials()
```

Comment:
action performed is EDT by default, so this needs to be moved to background

Reply:
good call - what assumptions can we make about : https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CorrectThreadCredentialsProvider.kt ?

Reply:
That is defensive.... if the call is on UI, and leads to an MFA promot it will deadlock the UI >.>

Reply:
The above should probably be switched to log.error only if wrong to prevent the dead log

Reply:
> The above should probably be switched to log.error only if wrong to prevent the dead lock

I don't understand what you mean by this.

Moved the other call off onto a pooled thread.

Reply:
CorrectThreadCredProvider should let the call go through on the UI thread I THINK...it may still deadlock with MFA (due to going into the SDK and back out).

Reply:
The above is not related to this PR, just an observation because you brought it up made me think about it.